File: Leetcode/Easy/1331/1331.RankTransformAnArray.cpp
// 1331. Rank Transform of An Array
#include <bits/stdc++.h>
using namespace std;

#pragma GCC optimize("Ofast")
#pragma GCC target("avx,avx2,fma")

typedef vector<int> vi;
typedef set<int> si;
#define ALL(x) x.begin(), x.end()

class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        if (arr.empty()) return {};
        
        vi unq(arr);
        sort(ALL(unq));
        auto duplicates = unique(ALL(unq));
        int uniqueElems = distance(unq.begin(),duplicates);
        if (uniqueElems == 1) return vi(size(arr),1);

        unordered_map<int,int> ranks;
        for (int i=0; i<uniqueElems; i++) ranks[unq[i]] = i+1;

        for (int i=0; i<size(arr); i++) arr[i] = ranks[arr[i]];
        return arr;
    }
};

File: Leetcode/Easy/1331.RankTransformAnArray.py
# 1331. Rank Transform of An Array
class Solution:
    def arrayRankTransform(self, arr: list[int]) -> list[int]:
        if not arr: return []
        
        unique = set(arr)
        if len(unique) == 1: return [1]*len(arr)

        ranks = {}
        unique = sorted(unique)
        for i in range(len(unique)):
            ranks[unique[i]] = i+1
        
        for i in range(len(arr)):
            arr[i] = ranks[arr[i]]
        return arr


# print(Solution.arrayRankTransform(None,[37,12,28,9,100,56,80,5,12]))
